pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		klog.ErrorS(err, "couldn't list namespaces")
+		return
+	}
+	podsForEviction := make(map[*v1.Pod]struct{})
+	// 1. for each namespace...
+	for _, namespace := range namespaces.Items {
+		namespacePods, err := client.CoreV1().Pods(namespace.Name).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			klog.ErrorS(err, "couldn't list pods in namespace", "namespace", namespace)
+			return
+		}
+
+		// ...where there is a topology constraint
+		for _, pod := range namespacePods.Items {
+			if pod.Spec.TopologySpreadConstraints != nil {
+				namespacedTopologySpreadConstrainPods[pod.Namespace] = append(namespacedTopologySpreadConstrainPods[pod.Namespace], pod)
+			}
+		}
+		if len(namespacedTopologySpreadConstrainPods[namespace.Name]) == 0 {
+			continue
+		}
+
+		// 2. for each topologySpreadConstraint in that namespace
+		for _, constrainedPod := range namespacedTopologySpreadConstrainPods[namespace.Name] {
+			// find pods to evict one constraint at a time
+			for _, constraint := range constrainedPod.Spec.TopologySpreadConstraints {
+				constraintTopologies := make(map[topologyPair][]*v1.Pod)
+
+				// 3. for each evictable pod in that namespace
+				// (this loop is where we count the number of pods per topologyValue that match this constraint's selector)
+				var sumPods float64
+				for _, pod := range namespacePods.Items {
+					if !evictable.IsEvictable(&pod) {
+						continue
+					}
+
+					// 4. if the pod matches this TopologySpreadConstraint LabelSelector
+					s, err := metav1.LabelSelectorAsSelector(constraint.LabelSelector)
+					if err != nil {
+						klog.ErrorS(err, "couldn't parse label selector as selector", "selector", constraint.LabelSelector)
+					}
+					if !s.Matches(labels.Set(pod.Labels)) {
+						continue
+					}
+
+					// 5. If the pod's node matches this constraint's topologyKey, create a topoPair and add the pod
+					node, err := client.CoreV1().Nodes().Get(ctx, pod.Spec.NodeName, metav1.GetOptions{})
+					if err != nil {
+						klog.ErrorS(err, "couldn't get node", "node", pod.Spec.NodeName)
+					}
+					nodeValue, ok := node.Labels[constraint.TopologyKey]
+					if !ok {
+						continue
+					}
+					// 6. create a topoPair with key as this TopologySpreadConstraint
+					topoPair := topologyPair{key: constraint.TopologyKey, value: nodeValue}
+					// 7. add the pod with key as this topoPair
+					constraintTopologies[topoPair] = append(constraintTopologies[topoPair], &pod)
+					sumPods++
+				}
+
+				// To actually determine how many pods need to be moved, we sort the topologies in ascending length
+				// Then, split the topologies in half around the ideal average (ie, perfect spreading)
+				// Sort the list of topologies with length <= the average in ascending order, and all > the average descending.
+				// For example, the list of topologies with these sizes (avg=5):
+				//  2
+				//  5
+				//  3
+				//  8
+				//  5
+				//  7
+				// Would end up in 2 lists like:
```

Comment:
Can you use list notation here? [2, 3, 5, 5] and [8, 7] to avoid quick confusion. Those reading lists by lines (including me) see 4 lists actually.